_posts/2019-02-05-configurable-builds-part-1.md

```diff
@@ -0,0 +1,280 @@
+   not building C++.
+1. `--crosstool_top`, which specifies exactly what command line should compile
+   your code, is even more tied to C++. Other languages may be organized
+   completely differently.
+1. `--define` is completely unstructured and unreadable in Starlark.
+1. `$ bazel build //my:all` sets the same flags for all targets, even if
+   different targets need different flags.
+1. `//my:binary`'s deps must use the same settings as their parent.
+1. It might be hard to remember which flags are needed by which targets,
+   especially across users and dev/test/prod environments. This risks
+   inconsistent and possibly incorrect builds.
+
+A different approach might be to set flags directly in targets that need them,
+let rule writers design the flags that affect their rules, and provide standard
+extensible APIs for cross-language concepts like *platform* and *cpu*. This
+lets people configure their projects on their own terms, using language
+consistent with how their projects are organized.
+
+Bazel's configurability work is redesigning Bazel around these ideas. It's
+ultimate goal is to make
+
+    $ bazel build //:all
+
+*just work*, which means every target knows how to build "the right way", for
+whatever platform(s) and features you care about, and you only have to set
+flags that are genuinely interesting to you. All while keeping builds fast and
+correct.
+
+## Configuration vs. Attributes
+
+Rules have always had [attributes](), which also affect how they build. The
+difference between configuration and attributes is that attributes only affect
+the rule's [direct build actions](https://docs.bazel.build/versions/master/skylark/rules.html#implementation-function)).
+This means attributes **cannot affect a rule's dependencies**.
+
+For example, C++ rules have an attribute named
+[`copts`](https://docs.bazel.build/versions/master/be/c-cpp.html#cc_binary.copts)
+that sets custom C++ compile options. `cc_binary(name = "mybinary", srcs =
+["mybinary.cc"], deps = [":mylibrary"], copts =
+["-DUSE_EXPERIMENTAL_FEATURES=1"])` might compile `mybinary.cc` with
+experimental features. But this won't happen for `:mylibrary` unless it sets
+its `copts` similarly. This makes attributes unsuitable for tasks like building
+a binary for a different CPU.
+
+If you imagine a build as a [graph](https://docs.bazel.build/versions/master/build-ref.html#dependencies)
+with parents above their dependencies, attributes change behavior
+[*up*](https://docs.bazel.build/versions/master/skylark/rules.html#providers)
+the graph while configuration changes behavior *down* the graph.
+
+## Platforms
+
+One of configurability's major efforts is designing a principled API for
+defining *platforms* and *toolchains*.
+
+While there are many kinds of "ways" you might want to build your project, in
+practice most developers want the flexibility to target different devices,
+OSes, CPUs, and other machine properties. The general term for this is
```

Comment:
I would cut the first part of this sentence, which I found confusing. Instead, I would say "Most modern developers want the flexibility to target specific devices, OSes, and CPU architectures—along with other machine properties."

Reply:
I feel like I need to say something more.

What I'm trying to fundamentally express is that "configuration" goes beyond just platforms, but platforms are one of *the* major use cases that cover a huge percentage of what people care about. 

I don't want to imply platforms is the *only* part that matters. We also need a story for the parts that don't fit into that (which I get more into in Part 2). And how all these different pieces link together as a holistic whole.

Reply:
Still totally open for rephrasing and I get your point about the confusion. But I'd like to integrate the above ideas somehow.